### ISSUE TITLE
[EPIC] Update Job Timeout

### ISSUE THREAD

**Issue by remcowesterhoud** [camunda/zeebe#15020]
### Feature name

⚠️ We don't like to name this feature **extend job lock**. First of all, it's not only used for extending. With this feature we will also support reducing the timeout of a Job, which we see as a valid use case. This is also something that's supported in C7. So instead of extend, we chose to call it updating.

Second, we don't have a locking mechanism in C8 as we do in C7. Therefore we don't want **lock** to be part of the name. We have never referred to Jobs as being locked in C8 and introducing this term now could cause confusion for users. Instead, we decided to go with timeout. This a term that users are already familiar with when working with job activations. So instead of job lock, we chose to call it job timeout.

**Tl;dr** we chose to name the topic **Update Job Timeout** as opposed to **Job Extend Lock**.

### Summary

There are jobs which have an elastic timespan, meaning they can run for 5 minutes, but can be also 24+ hours. For example in some projects there can be a lot of long-running background processes (duration is from several seconds to multiple hours and sometimes days).

The problem is that the workers picking up the jobs do not know in advance how long the given process take thus they can't accurately estimate a timeout.

In order to determine if a job is still OK, even though it takes a while, the job needs to extend the lock or update the timeout threshold. Using extremely long timeouts is no good idea, because then it can not be determined if a job failed a job.

### What does it look like in the state?

When a Job Worker activates a Job it can specify a timeout for how long the Job should remain activated (default: 5 minutes). In Zeebe we call this the deadline of the Job.
In the state 2 important changes are made when this happens:
1. The Job is removed from the `JOB_ACTIVATABLE` ColumnFamily
    - This makes it so no other Job Workers can activate the Job
2. The Job is added to the `JOB_DEADLINES` ColumnFamily
    - This ColumnFamily keeps track of the timestamp on which a Job activation needs to expire. Once it expires it is put back into the `JOB_ACTIVATABLE` ColumnFamily
    - This ColumnFamily has a composite key. The first part is the timestamp at which the Job activation expires. The second part is the Job key

### What should we change?

We should introduce a new command to gRPC. The Job Worker can periodically emit this command as long as it's still working on a Job to override the deadline. This command would look something like this:

```proto
message UpdateJobTimeoutRequest {
  // the unique job identifier, as obtained from ActivateJobsResponse
  int64 jobKey = 1;
  // the duration of the new timeout, starting from the current moment
  string timeout = 2;
}

message UpdateJobTimeoutResponse {
}
```

Notable is that it takes a duration. This is **not** the duration with which we will **extend** the timeout. Instead this will be the new duration the **timeout is set to from the current time**. This allows users to not only extend the timeout of a Job, but also gives users the option to shorten the timeout.

No tenant information needs to be provided in this command, as we have the Job key available. With this we can already check if the user is authorised to access this Job.

A new command also means we need to add new intents to the `JobIntent`. These will be:
- `UPDATE_TIMEOUT`
    -  This command will look for the Job in the state. If it can't find it it will return a `NOT_FOUND` rejection. Otherwise it will write the `TIMEOUT_UPDATED` event which will modify the deadline in the state.
    - This command will also look for a deadline in the state. If it can't find it it will return an `INVALID_STATE` rejection
- `TIMEOUT_UPDATED`
    - This event will update the `JOB_DEADLINES` ColumnFamily. Since the key of this ColumnFamily is `deadline-jockey` we can't just update the existing one. Instead we will need to remove the existing entry for this Job, and insert a new value. For this we must have access to the old deadline. This is stored in state and should available through the `JobRecord`.

## Task Breakdown

```[tasklist]
### Broker
- [ ] https://github.com/camunda/zeebe/issues/15034
- [ ] https://github.com/camunda/zeebe/issues/15035
- [ ] https://github.com/camunda/zeebe/issues/15036
```

```[tasklist]
### Gateway
- [ ] https://github.com/camunda/zeebe/issues/15037
- [ ] https://github.com/camunda/zeebe/issues/15038
```

```[tasklist]
### Zeebe Process Test
- [ ] https://github.com/camunda/zeebe-process-test/issues/965
```

```[tasklist]
### Clients
- [ ] https://github.com/camunda/zeebe/issues/15120
- [ ] https://github.com/camunda/zeebe/issues/15366
- [ ] https://github.com/camunda/zeebe/issues/15382
```

```[tasklist]
### Documentation
- [ ] https://github.com/camunda/camunda-docs/issues/2862
```

## Other Teams

N/A

## Discussions and Decisions

>❓ Should the client provide a duration, or a timestamp to become the new deadline?
>   - Duration is more UX friendly, no issues with different timezones on client-side and broker-side.
>   - We could decide to calculate the deadline on the gateway-side instead of the broker-side. This way we can't use the `ActorClock` of the Broker.
>   - If we calculate it broker-side we have to fit the duration in the `JobRecord`. We can either add a new property for this, or (a)buse the `deadlineProperty`. Putting the duration in the `deadlineProperty` makes it so the data in this property isn't exactly what the name suggests, but it is the easier approach. It only impacts the processor for this command, after which it's immediately set to the proper deadline again.

We've decided to add a `timeoutProperty` on the `JobRecord` and clarify with javadoc what it is used for. This makes a clear distinction in the record to hopefully reduce confusion. It also aligns better with job activation, where we users also define a timeout, which is the duration when a job times out after activating.

## Links

- Product Hub (internal): https://github.com/camunda/product-hub/issues/1734
- C7 API: https://docs.camunda.org/rest/camunda-bpm-platform/7.19-SNAPSHOT/#tag/External-Task/operation/extendLock

**Comment by remcowesterhoud** [comment]
- @remcowesterhoud Clarify that we don't like the Job Extend Lock as it's not really a locking mechanism in intro
- @remcowesterhoud Change naming from deadline to timeout to keep it aligned with job activation
- @remcowesterhoud Change naming from Change to Update
- @remcowesterhoud Check what happens if we have the UPDATE_TIMEOUT on the log, and also a TIMEOUT already after this.
    - No issue here, TIMEOUT Processor checks if deadline actually expired ✅ 
- @remcowesterhoud If we can't find the deadline consider returning an `INVALID_STATE` (?), something other that a `NOT_FOUND`.
- @remcowesterhoud Add a timeoutProperty to the JobRecord

**Comment by korthout** [comment]
@remcowesterhoud I suggest changing the name to `Update Job Deadline`. While the user specifies a timeout, the engine uses it to determine a new job deadline and update it accordingly. I think it best fits the topic.

This would also help avoid confusion with Job Backoff (which in theory is also a timeout keeping the job not-activatable but during the failed state).

**Comment by remcowesterhoud** [comment]
> @remcowesterhoud I suggest changing the name to `Update Job Deadline`. While the user specifies a timeout, the engine uses it to determine a new job deadline and update it accordingly. I think it best fits the topic.
> 
> This would also help avoid confusion with Job Backoff (which in theory is also a timeout keeping the job not-activatable but during the failed state).

I just updated everything 😭 
But really, I'm not sure if it's better. When activating jobs users also specify a timeout, which in turn sets a deadline. Users always work using the term timeout. The only time they see deadline is in the activate job response, which is likely hidden behind a job worker. I'm not convinced naming it `UpdateJobDeadline` is better tbh.

**Comment by korthout** [comment]
Good point, it's also called [Timeouts](https://docs.camunda.io/docs/next/components/concepts/job-workers/#timeouts) in the docs 👍

**Comment by ana-vinogradova-camunda** [comment]
Progress update: currently working on the clients part.
Will provide another update when ready.

**Comment by ana-vinogradova-camunda** [comment]
All tasks are completed and merged, closing the Epic as completed.

**Comment by abbasadel** [comment]
Great work Ana
